packages/@sfx/dom-events-plugin/src/dom-events-plugin.ts

```diff
@@ -0,0 +1,140 @@
+import { Plugin, PluginRegistry, PluginMetadata } from '@sfx/core';
+/**
+ * The browser version of the events plugin designed for the SF-X product.
+ * This plugin is responsible for exposing methods that will allow other
+ * plugins to register, unregister, and dispatch events.
+ */
+export default class DomEventsPlugin implements Plugin {
+  get metadata(): PluginMetadata {
+    return {
+      name: 'dom_events_plugin',
+      depends: [],
+    };
+  }
+
+  /**
+   * The core property is meant to hold a reference to the plugin registry
+   * that is passed in during the plugin registration lifecycle event.
+   */
+  core: PluginRegistry;
+  /**
+   * The exposedValue property is the value that the Events Browser Plugin
+   * exposes to the Core entity.
+   */
+  exposedValue: DomEventsPluginExposedValue;
+  /**
+   * The options property is the set of configuration values used to set
+   * up the plugin during construction.
+   */
+  options: DomEventsPluginOptions = {
+    window: typeof window !== "undefined" ? window : undefined,
+    CustomEvent: typeof CustomEvent !== "undefined" ? CustomEvent : undefined,
+  };
+  window: Window;
+  CustomEvent: typeof CustomEvent;
+
+  /**
+   * The browser events plugin constructor function which will combine
+   * both the default options and any options passed in. Also the constructor
+   * will accept a window object and a CustomEvent constructor provided
+   * through the options object.
+   *
+   * If either the window object or CustomEvent constructor provided are
+   * not valid, the plugin constructor will throw an error.
+   *
+   * @param options an object that must contain a window property that is
+   * a reference to the browser Window object.
+   */
+  constructor(options: Partial<DomEventsPluginOptions> = {}) {
+    this.options = {...this.options, ...options};
+    this.window = this.options.window;
+    this.CustomEvent = this.options.CustomEvent;
+
+    if (!this.window) {
+      throw new Error('window object is not valid');
+    }
+
+    if (typeof this.CustomEvent !== 'function') {
+      throw new Error('CustomEvent constructor is not valid');
+    }
+
+    // Binds
+    this.registerListener = this.registerListener.bind(this);
+    this.unregisterListener = this.unregisterListener.bind(this);
+    this.dispatchEvent = this.dispatchEvent.bind(this);
+  }
+
+  /**
+   * The register method that accepts a plugin registry object and returns
+   * the exposedValue of the browser events plugin. The exposedValue is an
+   * object that holds references to the registerListener, unregisterListener,
+   * and dispatchEvent methods.
+   *
+   * @param plugins a plugin registry object.
+   */
+  register(plugins: PluginRegistry): DomEventsPluginExposedValue {
+    this.core = plugins;
+
+    this.exposedValue = {
+      registerListener: this.registerListener,
+      unregisterListener: this.unregisterListener,
+      dispatchEvent: this.dispatchEvent,
+    };
```

Comment:
Can setting `this.exposedValue` be done in the constructor? If that were the case, we could actually rely on the `register` method from a super class. The super class could perform:

```js
this.core = plugins;
return this.exposedValue;
```

This would remove some boilerplate from all of our plugins. The only expectation would be to set exposedValue before register (ie. in the constructor), or override the parent's `register()` method.